src/modules/universal-discovery/universal.discovery.module.js

```diff
@@ -250,22 +286,53 @@ export default class UniversalDiscoveryModule extends Component {
     }
 
     /**
-     * Returns true or false if locations bookmark data was cached
-     * if not, checks it and then caches it
+     * Removes or adds bookmark depending on if it exists or not
+     *
+     * @method toggleBookmark
+     * @param {Object} location
+     * @memberof UniversalDiscoveryModule
+     */
+    toggleBookmark(location) {
+        const locationId = location.id;
+        const isBookmarked = this.isBookmarked(locationId);
+
+        if (this.isBookmarkLoading(locationId)) {
+            return;
+        }
+
+        this.setBookmarkLoadingStatus(locationId, true);
+
+        const { restInfo } = this.props;
+        const toggleBookmark = isBookmarked ? removeBookmark : addBookmark;
+        const onBookmarkToggled = isBookmarked ? this.onBookmarkRemoved : this.onBookmarkAdded;
+        const bookmarkToggled = new Promise((resolve) => toggleBookmark(restInfo, locationId, resolve));
+
+        bookmarkToggled
+            .then(() => {
+                onBookmarkToggled(location);
+                this.setBookmarkLoadingStatus(locationId, false);
+            })
+            .catch(showErrorNotification);
+    }
+
+    /**
+     * Returns
      *
      * @method isBookmarked
      * @param {String} locationId
+     * @returns {Boolean}
      * @memberof UniversalDiscoveryModule
      */
     isBookmarked(locationId) {
         const { bookmarked } = this.state;
-        if (locationId in bookmarked) {
-            return bookmarked[locationId];
-        }
+        const locationBookmarkChecked = locationId in bookmarked;
 
-        this.checkIsBookmarked(locationId);
+        if (!locationBookmarkChecked) {
+            this.checkIsBookmarked(locationId);
+            return null;
```

Comment:
And method `isBookmarked` returns `null` it's like: "I don't know" ;)

Reply:
Because it returns `null` to say "I don't know". :P
When the response from server has not yet arrived, you do not know if it is bookmarked or not. ;)

Reply:
IMO, there should be a promise, that updates state of bookmarked flag at the end. If we don't know, then it should return `false`.

Reply:
Otherwise, we'll end up like video detection in the browsers. `Maybe we support this video format, but who knows?`

Reply:
I like the promise to wait for the answer than returning "I don't know" or false.

Reply:
I don't like this approach. Elements should be rendered based on values from _state_ and _props_. This is how the cycle should look like in React. 
I don't know where you would like to put this Promise and how it would suppose to work. 
Maybe I don't understand what you are suggesting. :)

Reply:
I mean, default state is `false`. If content is bookmarked, then it will be changed to `true` when promise resolves.